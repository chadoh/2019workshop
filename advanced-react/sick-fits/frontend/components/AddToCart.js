import React, { Fragment } from 'react'
import PropTypes from 'prop-types'
import { Mutation } from 'react-apollo'
import gql from 'graphql-tag'
import Error from './ErrorMessage'
import { CURRENT_USER_QUERY } from './User'

const ADD_TO_CART_MUTATION = gql`
  mutation ADD_TO_CART_MUTATION($id: ID!) {
    addToCart(id: $id) {
      id
      quantity
    }
  }
`

const AddToCart = ({ id }) => (
  <Mutation
    mutation={ADD_TO_CART_MUTATION}
    variables={{ id }}
    refetchQueries={[ { query: CURRENT_USER_QUERY } ]}
  >
    {(addToCart, { error, loading }) => (
      <Fragment>
        <Error error={error} />
        <button
          onClick={addToCart}
          disabled={loading}
        >
          Add{loading && 'ing'} To Cart
        </button>
      </Fragment>
    )}
  </Mutation>
)


AddToCart.propTypes = {
  id: PropTypes.string.isRequired,
}

export default AddToCart
