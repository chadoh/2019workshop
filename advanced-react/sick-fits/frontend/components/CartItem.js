import React from 'react'
import PropTypes from 'prop-types'
import RemoveFromCart from './RemoveFromCart'
import formatMoney from '../lib/formatMoney'
import styled from 'styled-components'

const CartItemStyles = styled.li`
  padding: 1rem 0;
  border-bottom: 1px solid ${props => props.theme.lightgrey};
  display: grid;
  align-items: center;
  grid-template-columns: auto 1fr auto;
  grid-gap: 1em;
  img {
    width: 7em;
  }
  h3, p {
    margin: 0;
  }
`

const CartItem = ({ id, quantity, item }) => {
  return (
    <CartItemStyles>
      <img src={item.image} alt="" />
      <div className="cart-item-details">
        <h3>{item.title}</h3>
        <p>
          {formatMoney(item.price * quantity)}
          {' â€“ '}
          <em>
            {quantity} &times; {formatMoney(item.price)} each
          </em>
        </p>
      </div>
      <RemoveFromCart id={id} />
    </CartItemStyles>
  )
}

CartItem.propTypes = {
  id: PropTypes.string.isRequired,
  quantity: PropTypes.number.isRequired,
  item: PropTypes.shape({
    id: PropTypes.string.isRequired,
    title: PropTypes.string.isRequired,
    image: PropTypes.string.isRequired,
    price: PropTypes.number.isRequired,
  }).isRequired,
}

export default CartItem
