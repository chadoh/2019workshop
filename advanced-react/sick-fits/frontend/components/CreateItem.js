import React, { Component } from 'react'
import { Mutation } from 'react-apollo'
import gql from 'graphql-tag'
import Router from 'next/router'
import Form from './styles/Form'
import formatMoney from '../lib/formatMoney'
import Error from './ErrorMessage'

export const CREATE_ITEM_MUTATION = gql`
mutation CREATE_ITEM_MUTATION(
  $title: String!
  $price: Int!
  $description: String!
  $image: String
  $largeImage: String
) {
  createItem(
    title: $title
    price: $price
    description: $description
    image: $image
    largeImage: $largeImage
  ) {
    id
  }
}
`

class CreateItem extends Component {
  state = {
    title: '',
    description: '',
    image: '',
    largeImage: '',
    price: '',
  }

  handleChange = (e) => {
    const { name, type, value } = e.target
    const val = type === 'number' ? parseFloat(value) : value
    this.setState({ [name]: val })
  }

  render() {
    return (
      <Mutation
        mutation={CREATE_ITEM_MUTATION}
        variables={this.state}
      >
        {(createItem, { loading, error }) => (
          <Form
            onSubmit={async e => {
              e.preventDefault()
              const res = await createItem()
              Router.push({
                pathname: '/item',
                query: { id: res.data.createItem.id },
              })
            }}
          >
            <Error error={error} />
            <fieldset disabled={loading} aria-busy={loading}>
              <label htmlFor="title">
                Title
                <input
                  type="text"
                  id="title"
                  name="title"
                  required
                  autoComplete="off"
                  value={this.state.title}
                  onChange={this.handleChange}
                />
              </label>
              <label htmlFor="price">
                Price
                <input
                  type="number"
                  id="price"
                  name="price"
                  required
                  autoComplete="off"
                  value={this.state.price}
                  onChange={this.handleChange}
                />
              </label>
              <label htmlFor="description">
                Description
                <input
                  type="text"
                  id="description"
                  name="description"
                  required
                  autoComplete="off"
                  value={this.state.description}
                  onChange={this.handleChange}
                />
              </label>
              <button type="submit">Submit</button>
            </fieldset>
          </Form>
        )}
      </Mutation>
    )
  }
}

export default CreateItem
