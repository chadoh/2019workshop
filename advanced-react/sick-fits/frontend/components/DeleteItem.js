import React from 'react'
import PropTypes from 'prop-types'
import { Mutation } from 'react-apollo'
import gql from 'graphql-tag'
import { ALL_ITEMS_QUERY } from './Items'
import { CURRENT_USER_QUERY } from './User'

const DELETE_ITEM_MUTATION = gql`
  mutation DELETE_ITEM_MUTATION($id: ID!) {
    deleteItem(id: $id) {
      id
    }
  }
`

const DeleteItem = ({ id }) => (
  <Mutation
    mutation={DELETE_ITEM_MUTATION}
    variables={{ id }}
    refetchQueries={[
      { query: CURRENT_USER_QUERY },
      { query: ALL_ITEMS_QUERY },
    ]}
  >
    {(deleteItem, { error, loading }) => {
      return (
        <button
          onClick={() => {
            if (confirm('Are you sure you want to delete this item?')) {
              deleteItem().catch(err => {
                alert(err.message.replace('GraphQL error: ', ''))
              })
            }
          }}
        >
          Delete Item
        </button>
      )
    }}
  </Mutation>
)

DeleteItem.propTypes = {
  id: PropTypes.string.isRequired,
}

export default DeleteItem
