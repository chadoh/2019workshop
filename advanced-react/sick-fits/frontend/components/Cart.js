import React from 'react'
import { Query, Mutation } from 'react-apollo'
import gql from 'graphql-tag'
import PropTypes from 'prop-types'
import User from './User'
import CartStyles from './styles/CartStyles'
import Supreme from './styles/Supreme'
import CloseButton from './styles/CloseButton'
import SickButton from './styles/SickButton'
import CartItem from './CartItem'
import formatMoney from '../lib/formatMoney'
import calcTotalPrice from '../lib/calcTotalPrice'

export const LOCAL_STATE_QUERY = gql`
  query LOCAL_STATE_QUERY {
    cartOpen @client
  }
`

export const TOGGLE_CART_MUTATION = gql`
  mutation TOGGLE_CART_MUTATION {
    toggleCart @client
  }
`

const Cart = ({  }) => (
  <User>
    {({ data: { me} }) => {
      if (!me) return null
      return (
        <Mutation mutation={TOGGLE_CART_MUTATION}>
          {toggleCart => (
            <Query query={LOCAL_STATE_QUERY}>
              {({ data }) => (
                <CartStyles open={data.cartOpen}>
                  <header>
                    <CloseButton
                      title="close"
                      onClick={toggleCart}
                    >
                      &times;
                    </CloseButton>
                    <Supreme>Your Cart</Supreme>
                    <p>You have {me.cart.length} Item{me.cart.length === 1 ? '' : 's'} in your cart.</p>
                  </header>
                  <ul>
                    {me.cart.map(cartItem => (
                      <CartItem key={cartItem.id} {...cartItem} />
                    ))}
                  </ul>
                  <footer>
                    <p>{formatMoney(calcTotalPrice(me.cart))}</p>
                    <SickButton>Checkout</SickButton>
                  </footer>
                </CartStyles>
              )}
            </Query>
          )}
        </Mutation>
      )
    }}
  </User>
)


Cart.propTypes = {
}

export default Cart
