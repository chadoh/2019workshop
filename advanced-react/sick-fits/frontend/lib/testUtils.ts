import casual from 'casual'
import { ApolloClient } from 'apollo-client'
import { CurrentUserResponse } from '../__tests__/SignUp.test'
import { CartItemInterface } from '../components/CartItem'
import { CURRENT_USER_QUERY } from '../components/User'

// seed it so we get consistent results
casual.seed(777)

const fakeItem = () => ({
  __typename: 'Item',
  id: 'abc123',
  price: 5000,
  user: null,
  image: 'dog-small.jpg',
  title: 'dogs are best',
  description: 'dogs',
  largeImage: 'dog.jpg',
})

const fakeUser = () => ({
  __typename: 'User',
  id: '4234',
  name: casual.name,
  email: casual.email,
  permissions: ['ADMIN'],
  cart: [],
})

const fakeOrderItem = () => ({
  __typename: 'OrderItem',
  id: casual.uuid,
  image: `${casual.word}.jpg`,
  title: casual.words(),
  price: 4234,
  quantity: 1,
  description: casual.words(),
})

const fakeOrder = () => ({
  __typename: 'Order',
  id: 'ord123',
  charge: 'ch_123',
  total: 40000,
  items: [fakeOrderItem(), fakeOrderItem()],
  createdAt: '2018-04 - 06T19: 24: 16.000Z',
  user: fakeUser(),
})

const fakeCartItem = (overrides?: object): object => ({
  __typename: 'CartItem',
  id: 'omg123',
  quantity: 3,
  item: fakeItem(),
  user: fakeUser(),
  ...overrides,
})

// Fake LocalStorage
class LocalStorageMock {
  store: { [key: string]: any };

  constructor() {
    this.store = {}
  }

  clear() {
    this.store = {}
  }

  getItem(key: string) {
    return this.store[key] || null
  }

  setItem(key: string, value: any) {
    this.store[key] = value.toString()
  }

  removeItem(key: string) {
    delete this.store[key]
  }
}

export const refetchCart = async (
  client: ApolloClient<any>
): Promise<CartItemInterface[]> => {
  const res = await client.query<CurrentUserResponse>({
    query: CURRENT_USER_QUERY,
  })
  return res.data.me.cart
}

export {
  LocalStorageMock,
  fakeItem,
  fakeUser,
  fakeCartItem,
  fakeOrder,
  fakeOrderItem,
}
