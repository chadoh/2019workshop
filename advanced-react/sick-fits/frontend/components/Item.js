import React from 'react';
import PropTypes from 'prop-types';
import Link from 'next/link'
import DeleteItem from './DeleteItem';
import Title from './styles/Title';
import ItemStyles from './styles/ItemStyles';
import PriceTag from './styles/PriceTag';
import formatMoney from '../lib/formatMoney';
import AddToCart from './AddToCart'

function Item({item}) {
  return <ItemStyles>
    {item.image && <img src={item.image} alt="" />}
    <Title>
      <Link href={{
        pathname: '/item',
        query: { id: item.id },
      }}>
        <a>{item.title}</a>
      </Link>
    </Title>
    <PriceTag>{formatMoney(item.price)}</PriceTag>
    <p>{item.description}</p>

    <div className="buttonList">
      <Link href={{
        pathname: 'update',
        query: { id: item.id },
      }}>
        <a>Edit ✏️</a>
      </Link>
      <AddToCart id={item.id} />
      <DeleteItem id={item.id} />
    </div>
  </ItemStyles>;
}

Item.propTypes = {
  item: PropTypes.object.isRequired,
};

export default Item;
