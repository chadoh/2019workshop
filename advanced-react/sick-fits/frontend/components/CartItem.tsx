import React from 'react'
import styled from 'styled-components'
import RemoveFromCart from './RemoveFromCart'
import { ItemInterface } from './Item'
import formatMoney from '../lib/formatMoney'

export interface CartItemInterface {
  id: string;
  quantity: number;
  item: ItemInterface;
}

const CartItemStyles = styled.li`
  padding: 1rem 0;
  border-bottom: 1px solid ${props => props.theme.lightgrey};
  display: grid;
  align-items: center;
  grid-template-columns: auto 1fr auto;
  grid-gap: 1em;
  img {
    width: 7em;
  }
  h3,
  p {
    margin: 0;
  }
`

const CartItem: React.SFC<CartItemInterface> = ({ id, quantity, item }) => {
  return (
    <CartItemStyles>
      <img src={item.image} alt="" />
      <div className="cart-item-details">
        <h3>{item.title}</h3>
        <p>
          {formatMoney(item.price * quantity)}
          {' â€“ '}
          <em>
            {quantity} &times; {formatMoney(item.price)} each
          </em>
        </p>
      </div>
      <RemoveFromCart id={id} />
    </CartItemStyles>
  )
}
export default CartItem
