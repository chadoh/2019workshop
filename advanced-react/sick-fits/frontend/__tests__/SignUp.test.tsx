import { mount, ReactWrapper } from 'enzyme'
import wait from 'waait'
import toJSON from 'enzyme-to-json'
import { ApolloClient } from 'apollo-client'
import { MockedProvider } from 'react-apollo/test-utils'
import { ApolloConsumer } from 'react-apollo'
import SignUp, { SIGNUP_MUTATION } from '../components/SignUp'
import { CURRENT_USER_QUERY } from '../components/User'
import { fakeUser } from '../lib/testUtils'
import { CartItemInterface } from '../components/CartItem'

const type = (wrapper: ReactWrapper, name: string, value: string) => {
  wrapper.find(`input[name="${name}"]`).simulate('change', {
    target: { name, value },
  })
}

const me = fakeUser()
const password = 'cool'

const signupMock = {
  request: {
    query: SIGNUP_MUTATION,
    variables: {
      name: me.name,
      email: me.email,
      password,
    },
  },
  result: {
    data: {
      signup: {
        __typename: 'User',
        id: me.id,
        name: me.name,
        email: me.email,
      },
    },
  },
}

const currentUserMock = {
  request: { query: CURRENT_USER_QUERY },
  result: { data: { me } },
}

export interface UserInterface {
  id: string;
  name: string;
  email: string;
  cart: CartItemInterface[];
}

interface CurrentUserResponse {
  me: UserInterface;
}

describe('<SignUp/>', () => {
  it('renders and matches snapshot', async () => {
    const wrapper = mount(
      <MockedProvider>
        <SignUp />
      </MockedProvider>
    )
    expect(toJSON(wrapper.find('form'))).toMatchSnapshot()
  })

  it('calls the mutation properly', async () => {
    let apolloClient: ApolloClient<any> | undefined
    const wrapper = mount(
      <MockedProvider mocks={[signupMock, currentUserMock]}>
        <ApolloConsumer>
          {client => {
            apolloClient = client
            return <SignUp />
          }}
        </ApolloConsumer>
      </MockedProvider>
    )
    await wait()
    wrapper.update()
    type(wrapper, 'name', me.name)
    type(wrapper, 'email', me.email)
    type(wrapper, 'password', password)
    wrapper.update()
    wrapper.find('form').simulate('submit')
    await wait()
    if (!apolloClient) throw new Error('apolloClient never set')
    const user = await apolloClient.query<CurrentUserResponse>({
      query: CURRENT_USER_QUERY,
    })
    expect(user.data.me).toMatchObject(me)
  })
})
