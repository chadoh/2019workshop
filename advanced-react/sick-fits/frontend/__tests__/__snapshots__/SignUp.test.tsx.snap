// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SignUp/> renders and matches snapshot 1`] = `
<form
  className="Form-sc-1xszr8q-0 krRxky"
  method="post"
  onSubmit={[Function]}
>
  <fieldset
    aria-busy={false}
    disabled={false}
  >
    <h2>
      Sign up for an account
    </h2>
    <DisplayError
      error={Object {}}
    />
    <label
      htmlFor="email"
    >
      Email
      <input
        autoComplete="off"
        name="email"
        onChange={[Function]}
        required={true}
        type="email"
        value=""
      />
    </label>
    <label
      htmlFor="name"
    >
      Name
      <input
        autoComplete="off"
        name="name"
        onChange={[Function]}
        required={true}
        type="text"
        value=""
      />
    </label>
    <label
      htmlFor="password"
    >
      Password
      <input
        autoComplete="off"
        name="password"
        onChange={[Function]}
        required={true}
        type="password"
        value=""
      />
    </label>
    <button
      type="submit"
    >
      Sign Up
    </button>
  </fieldset>
</form>
`;
